### PR TITLE
HRINT-1576

### PR DESCRIPTION
**Pull request by ppekrol**
- added limits to RequestExecutor context pool
- switch from Queue to Stack to constantly reuse same contexts instead of reusing all of them, this should allow to clear them more easily

### CODE REVIEW

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 1]
```diff
@@ -352,7 +353,15 @@ protected RequestExecutor(string databaseName, X509Certificate2 certificate, Doc
 
             _lastReturnedResponse = DateTime.UtcNow;
 
-            ContextPool = new JsonContextPool(new Size(8, SizeUnit.Megabytes));
+            var maxContextSizeToKeep = PlatformDetails.Is32Bits == false
```
Comment (grisha-kotler) [inline src/Raven.Client/Http/RequestExecutor.cs]: Isn't it too small for 64bit?
Can we check what is the average size that we have?

Comment (ayende) [reply src/Raven.Client/Http/RequestExecutor.cs]: Agreed, and I think we want to make this configurable for the user.

File: src/Sparrow/Json/JsonContextPoolBase.cs [hunk 2]
```diff
@@ -152,57 +182,95 @@ public void Dispose()
...
-                if (currentTime - context.InPoolSince < idleTime)
+                if (globalStackRebuildNeeded)
                 {
-                    break;
+                    _lastGlobalStackRebuild = currentTime;
+
+                    var previousGlobalStack = _previousGlobalStack;
+
+                    _currentGlobalStack = new CountingConcurrentStack<T>();
+                    _previousGlobalStack = currentGlobalStack;
```
Comment (ayende) [inline src/Sparrow/Json/JsonContextPoolBase.cs]: Isn't this a problem when we are running concurrently?
Other threads may add to their view of `_currentGlobalStack` after you called `ClearStack`

Comment (ayende) [reply src/Sparrow/Json/JsonContextPoolBase.cs]: And why not simply clearing the *current* stack and be done with it?

File: src/Sparrow/Json/JsonContextPoolBase.cs [hunk 3]
```diff
@@ -1,24 +1,30 @@
...
         private bool _disposed;
 
         protected SharedMultipleUseFlag LowMemoryFlag = new SharedMultipleUseFlag();
         private readonly CancellationTokenSource _cts = new CancellationTokenSource();
         private readonly long _maxContextSizeToKeepInBytes;
+        private readonly long _maxNumberOfContextsToKeepInGlobalStack;
 
         private readonly T[][] _perCoreContexts;
-        private readonly ConcurrentQueue<T> _globalQueue = new ConcurrentQueue<T>();
+        private CountingConcurrentStack<T> _previousGlobalStack;
```
Comment (ayende) [inline src/Sparrow/Json/JsonContextPoolBase.cs]: Why the change from queue to stack?

File: src/Raven.Client/Http/RequestExecutor.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Client/Http/RequestExecutor.cs]: My thinking is:
a) this is Client API, so why it should retain big context at all?
b) this is Client API, so why it should retain high amount of contexts?

I've checked our api.ravendb.net, most of the context are below 1MB, with few that are up to 4MB of size, with total 4GB of memory kept in context pool.

File: src/Sparrow/Json/JsonContextPoolBase.cs (hunk 3 above)
Comment (ppekrol) [reply src/Sparrow/Json/JsonContextPoolBase.cs]: If we have a queue with 4000 items in it already and we are requesting let's say 20 contexts per second (synchronously even - no concurrency). It means that in 1min we will take 1200 different contexts from the pool and refresh their 'last used date', so the cleaner will not trigger, but by using stack, we will use only 1 context from the stack, so others can be safely removed by the cleaner.

File: src/Sparrow/Json/JsonContextPoolBase.cs (hunk 2 above)
Comment (ppekrol) [reply src/Sparrow/Json/JsonContextPoolBase.cs]: The idea here is that I still want to retain (maybe unnecessary?) contexts from the global stack that are not outdated. I'm doing that by creating new stack, but when allocating I'm taking values first from the 'old stack', then releasing them to the 'current stack'. I'm doing that, because I'm not removing anything from the stack during the cleanup (regular one - not the one that happens every 15min), only disposing contexts there.

The other option would be just to release everything as you wrote. It is easier, but not sure if we want that?

File: src/Sparrow/Json/JsonContextPoolBase.cs [hunk 4]
```diff
@@ -1,25 +1,32 @@
...
         private readonly CancellationTokenSource _cts = new CancellationTokenSource();
         private readonly long _maxContextSizeToKeepInBytes;
+        private readonly long _maxNumberOfContextsToKeepInGlobalStack;
+        private long _numberOfContextsDisposedInGlobalStack;
 
         private readonly T[][] _perCoreContexts;
-        private readonly ConcurrentQueue<T> _globalQueue = new ConcurrentQueue<T>();
-        private readonly Timer _idleTimer;
+        private CountingConcurrentStack<T> _previousGlobalStack;
+        private CountingConcurrentStack<T> _currentGlobalStack = new CountingConcurrentStack<T>();
```
Comment (ayende) [inline src/Sparrow/Json/JsonContextPoolBase.cs]: Why switch from queue to stack?

File: src/Sparrow/Json/JsonContextPoolBase.cs [hunk 5]
```diff
@@ -225,26 +298,34 @@ private void Push(T context)
...
+
             // couldn't find a place for it, let's add it to the global list
-            _globalQueue.Enqueue(context);
+            if (currentGlobalStack.Count >= _maxNumberOfContextsToKeepInGlobalStack)
+            {
+                context.Dispose();
+                return;
+            }
+
+            currentGlobalStack.Push(context);
```
Comment (ayende) [inline src/Sparrow/Json/JsonContextPoolBase.cs:305]: Here you may be pushing to the previous instance, no?

File: src/Sparrow/Json/JsonContextPoolBase.cs [hunk 6]
```diff
@@ -152,57 +183,99 @@ public void Dispose()
...
-                if (currentTime - context.InPoolSince < idleTime)
+                if (globalStackRebuildNeeded && _numberOfContextsDisposedInGlobalStack > 0)
                 {
-                    break;
+                    _lastGlobalStackRebuild = currentTime;
+
+                    var previousGlobalStack = _previousGlobalStack;
+
+                    _numberOfContextsDisposedInGlobalStack = 0;
+                    _currentGlobalStack = new CountingConcurrentStack<T>();
```
Comment (ayende) [inline src/Sparrow/Json/JsonContextPoolBase.cs]: Why are you doing this switch?

File: src/Raven.Client/Http/RequestExecutor.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Client/Http/RequestExecutor.cs]: This is also used internally for things like ETL, where we might have larger values.
Should we customize the conventions there?

File: src/Sparrow/Json/JsonContextPoolBase.cs (hunk 2 above)
Comment (ayende) [reply src/Sparrow/Json/JsonContextPoolBase.cs]: The easiest thing to do would be to keep the queue structure, and scan the whole queue.
If there are old stuff there, just mark it in use and dispose it.
The null pull from the queue will clear it up, no?

File: src/Sparrow/Json/JsonContextPoolBase.cs (hunk 3 above)
Comment (ayende) [reply src/Sparrow/Json/JsonContextPoolBase.cs]: IIRC, there was a reasoning behind this.
The idea is that we aren't going to end up with so many items in the global pool, because that is quite strange to have.
And when you pull from the global pool it goes to the per core one, no?
That would avoid this issue.

The auto cleanup of old instance really helps here.

File: src/Raven.Client/Http/RequestExecutor.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Client/Http/RequestExecutor.cs]: Increased to 8/2 MB (64/32 bit respectively) for server
